data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/PeerForwarder.java

```diff
@@ -0,0 +1,175 @@
+                    LOG.error("Problem with asynchronous peer forwarding", e);
+                }
+            }
+        }
+        return recordsToProcessLocally;
+    }
+
+    private Map<String, List<Record<Event>>> groupRecordsBasedOnIdentificationKeys(Collection<Record<Event>> records, Set<String> identificationKeys) {
+        final Map<String, List<Record<Event>>> groupedRecords = new HashMap<>();
+
+        // group records based on IP address calculated by HashRing
+        for (final Record<Event> record : records) {
+            final Event event = record.getData();
+
+            List<String> identificationKeyValues = new LinkedList<>();
+            for (final String identificationKey : identificationKeys) {
+                identificationKeyValues.add(event.get(identificationKey, Object.class).toString());
+            }
+
+            final String dataPrepperIp = hashRing.getServerIp(identificationKeyValues).orElse(StaticPeerListProvider.LOCAL_ENDPOINT);
+            groupedRecords.computeIfAbsent(dataPrepperIp, x -> new ArrayList<>()).add(record);
+        }
+        return groupedRecords;
+    }
+
+    private WebClient getClient(final String address) {
+        return isAddressDefinedLocally(address) ? LOCAL_CLIENT : peerClientPool.getClient(address);
+    }
+
+    private boolean isAddressDefinedLocally(final String address) {
+        final InetAddress inetAddress;
+        try {
+            inetAddress = InetAddress.getByName(address);
+        } catch (UnknownHostException e) {
+            return false;
+        }
+        if (inetAddress.isAnyLocalAddress() || inetAddress.isLoopbackAddress()) {
+            return true;
+        } else {
+            try {
+                return NetworkInterface.getByInetAddress(inetAddress) != null;
+            } catch (SocketException e) {
+                return false;
+            }
+        }
+    }
+
+    private boolean isLocalClient(final WebClient client) {
+        return client == LOCAL_CLIENT;
+    }
+
+    private String serializeRecordsToJsonString(final List<Record<Event>> records) throws JsonProcessingException {
+        List<SerializedEvent> serializedEvents = new ArrayList<>();
+
+        for (Record<Event> record: records) {
+            final Event event = record.getData();
+            serializedEvents.add(new SerializedEvent(event.getMetadata().getEventType(), event.toJsonString()));
+        }
+
+        // TODO: get plugin id from PipelineParser
+        final WireEvents wireEvents = new WireEvents(serializedEvents, "pluginId");
+
+        return objectMapper.writeValueAsString(wireEvents);
+    }
+
+    private HttpRequest createHttpRequest(final String content) {
+        return HttpRequest.builder()
+                .content(content)
+                .method(HttpMethod.POST)
+                .build();
+    }
+
+    private CompletableFuture<AggregatedHttpResponse> forwardRecords(final WebClient client, final HttpRequest httpRequest) {
```

Comment:
You can move all of the work that involves the client into a class like `PeerForwarderClient`. This class can be responsible for making the HTTP request using `WebClient`. It would also be the best class to use the serialization/deserialization class that I suggested in a comment above.

With this change, the `PeerForwarder` doesn't need to know anything about HTTP (it won't need a `WebClient`) or serialization.